### DISCUSSION TITLE
JavaScript Functions

### DISCUSSION THREAD

**Discussion by dianzrong** [zauberzeug/nicegui#940] [Q&A]
### Question

Hi,

Could I get some clarification on how to use a JavaScript function I defined in NiceGUI? Specifically, I'm trying to implement the [LoomSDK](https://dev.loom.com/docs/record-sdk/getting-started#implement).

Currently I have 3 files:
1. the LoomSDK implementation with 1 function in JavaScript
2. a subclass of element called Loom with an initializer which calls register_component (not optimal I know, but gets something to show on the screen) and the super's initializer and an object method that calls the function in the Javascript using self.run_method
3. the main file which calls run_javascrupt with await and the object function in the 2nd file

When I run the main file, the buttons appear and are clickable but the Loom display doesn't appear when the button is clicked.

**Comment by falkoschindler** [comment]
I think it's best if you show your code so we can help more efficiently.

**Reply by dianzrong** [reply]
updated_loom.js:
```js
import { isSupported, setup } from "@loomhq/record-sdk";
import { oembed } from "@loomhq/loom-embed";

const PUBLIC_APP_ID = "public-app-id-obtained-from-developer-portal";
const BUTTON_ID = "loom-record-sdk-button";

export default async function initialize_loom() {
  const { supported, error } = await isSupported();

  if (!supported) {
    console.warn(`Error setting up Loom: ${error}`);
    return;
  }

  const root = document.getElementById("app");

  if (!root) {
    return;
  }

  root.innerHTML = `<button id="${BUTTON_ID}">Record</button>`;

  const button = document.getElementById(BUTTON_ID);

  if (!button) {
    return;
  }

  const { configureButton } = await setup({
    publicAppId: PUBLIC_APP_ID,
  });

  const sdkButton = configureButton({ element: button });

  sdkButton.on("insert-click", async (video) => {
    const { html } = await oembed(video.sharedUrl, { width: 400 });
    const target = document.getElementById("target");

    if (target) {
    target.innerHTML = html;
  }
  });
}
```

Loom.py:
```py
from typing import Callable, Optional

from nicegui import ui
from nicegui.dependencies import register_component
from nicegui.element import Element

class Loom(Element):
    """
    The application where loom is used to record videos.
    """

    def __init__(self) -> None: 
        register_component('loom', __file__, 'updated_loom.js')
        super().__init__('loom')

    async def start_recording(self) -> None:
        """
        When the button is clicked, start the recording.
        """
        await self.run_method('initialize_loom')
```
share.py:
```py
import tracemalloc
from nicegui import ui
from nicegui.element import Element
from loom import Loom

async def start():
    recorder = Loom()
    ui.run_javascript('await recorder.start_recording()')  # probs change
    ui.notify('The recording has started')

async def get_date():
    time = await ui.run_javascript('Date()')
    ui.notify(f'Browser time: {time}')

ui.button('start recording', on_click=start)
ui.button('receive result', on_click=get_date)

ui.run(port=8000)
```

**Reply by falkoschindler** [reply]
A few things that I noticed while skimming your code:
- `register_component` should be called from the global scope at import time. See our [custom Vue component example](https://github.com/zauberzeug/nicegui/tree/main/examples/custom_vue_component).
- `ui.run_javascript` should be awaited.
- loom.js does not implement a Vue component, but a JavaScript function.

Unfortunately, I currently don't have the time to deep-dive into the implementation of a loom component. Nevertheless, I hope you manage to do so, as it would be a great feature. Maybe somebody else can help out?

For reference, you might want to look into existing custom components in the [elements](https://github.com/zauberzeug/nicegui/tree/main/nicegui/elements) directory (basically all *.js files).
